bridge-fungible-token/src/utils.sw

```diff
@@ -1,11 +1,184 @@
 library utils;
 
-use std::mem::read;
+dep errors;
+dep events;
+dep data;
+
+use std::{
+    alloc::alloc,
+    mem::copy,
+    outputs::{
+        Output,
+        output_count,
+        output_type,
+    },
+    revert::revert,
+    vec::Vec,
+};
+
+use errors::BridgeFungibleTokenError;
+use events::{BurnEvent, MintEvent, TransferEvent};
+use data::MessageData;
+use std::{
+    constants::ZERO_B256,
+    inputs::{
+        Input,
+        input_pointer,
+        input_type,
+    },
+    logging::log,
+    mem::{
+        addr_of,
+        read,
+        write,
+    },
+    token::{
+        burn,
+        mint_to_address,
+        transfer_to_address,
+    },
+    vm::evm::evm_address::EvmAddress,
+};
 
 // TODO: [std-lib] remove once standard library functions have been added
 const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
 const GTF_INPUT_MESSAGE_DATA = 0x11E;
+const GTF_INPUT_MESSAGE_SENDER = 0x115;
+const GTF_INPUT_MESSAGE_RECIPIENT = 0x116;
+
+pub fn mint_and_transfer_tokens(amount: u64, to: Address) {
+    mint_to_address(amount, to);
+    log(MintEvent { amount, to });
+}
+
+pub fn burn_tokens(amount: u64, from: Identity) {
+    burn(amount);
+    log(BurnEvent {
+        from: from,
+        amount,
+    })
+}
+
+pub fn correct_input_type(index: u64) -> bool {
+    let type = input_type(index);
+    match type {
+        Input::Message => {
+            true
+        },
+        _ => {
+            false
+        }
+    }
+}
+
+pub fn safe_b256_to_u64(val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    // first, decompose into u64 values
+    let u64s = decompose(val);
+    log(u64s.3);
+
+    // verify amount will require no partial refund of dust by ensuring that
```

Comment:
current max bridgeable amount == 18446744073000000000. This is the value of `u64::max()`, with the last 9 places zeroed.
I'll probably need to modify this to allow the full `u64::max` coins to be bridged.

Reply:
I'm not 100% clear on handling decimals.
The L1 erc20 contracts typically use 18 decimals, and this L2 contract uses 9 decimals.
Does this mean that 1_000_000_000 units (wei) on Ethereum, when bridged, becomes 1 unit on Fuel?
**i.e: 1 token is till 1 token**
Eth  =>   1.000_000_000_000_000_000 
Fuel =>  1.000_000_000

**but we drop 9 decimal places, so 1_000_000_000 of the smallest units on eth becomes 1 unit on Fuel**
Eth   => 0.000_000_001_000_000_000
Fuel => 0.000_000_001

Reply:
I think what this means is that when withdrawing, the amount encoded in the message data should be the number of token units (**n**) sent to the contract **adjusted** to get the decimals correct for ethereum, so **n * 1_000_000_000**

Reply:
cc @pixelcircuits

Reply:
Your thinking is correct, but you're thinking specifically in terms of how ETH is going to work on Fuel. So to reiterate, 1 ETH on Fuel will equal 1 ETH on Ethereum. Ethereum has 18 decimals of precision **_for ETH_**, but Fuel only has 9 _**for ETH**_. This means that you cannot bridge ETH amounts less than 0.00_000_001 from Ethereum to Fuel. You also cannot bridge an ETH amount larger than 18_446_744_074 (over 100x the current ETH supply, btw) since this is the limit of a u64 with 9 decimals of precision.

However, this contract has to fit most ERC20 tokens which may not be using 18 decimals of precision on the Ethereum side like ETH. I think most do use 18, but the ERC20 standard allows for any decimal value to be chosen. On top of that, this Fuel side contract may not want to use 9 decimals of precision, but again most probably will.

The decimals for both the L1 side (specified in the ERC20 contract) and the L2 side need to be defined somewhere. Then these u256 <-> u64 conversion utils need to know what those decimal levels are so they can do the appropriate conversion as well as check for both values too small or too large.

Reply:
Yep, this is all in line what how I'm thinking about it. The decimal conversion function will ideally use something dynamic, where the factor to multiply by would be 
`token_amount * (10 ** (LAYER_1_DECIMALS - LAYER_2_DECIMALS))` , assuming no need to ever use > 18 decimals on Fuel.

Reply:
The `**` syntax doesn't work atm, but isn't much work to implement. There is also an ongoing discussion in the vm repo about improving the accuracy of the `mlog` opcode.